### ISSUE TITLE
Edge WebView2 Runtime lets our Store Office Add In fail - too many redirects

### ISSUE THREAD

**Issue by ndeeH** [MicrosoftEdge/WebView2Feedback#1194]
Hi there,
we have an Office AddIn (Content Add In for PowerPoint) in the store for a long time.
(https://appsource.microsoft.com/de-de/product/office/WA104381814?src=office&tab=Overview)

With this Office Version we get a "Too many redirect" error when webview2 is installed while the Add In is authenticating:
![image](https://user-images.githubusercontent.com/11518242/115018601-5a3c3b80-9eb8-11eb-90aa-4d48014d1478.png)

The add in and the server we are connecting do not have changed - only office / windows was updated.
Now I simply **un-installed** the Microsoft Edge WebView2 Runtime (89.0.774.76)
Now our Add In works again like expected.

Any ideas or is this just a bug in WebView2?
Thank you, Andi

**Version**
Runtime: 89.0.774.76
Framework: Office Add In
OS: Win10 1909 18363.1500

**Repro Steps**
- Install Add In (Link in third line) 
- open Powerpoint and insert add in
- copy this url to clipboard:
https://preview.idl.eu/Portal/Apps/ReportViewer/Default.aspx?id=b3956ff7-c363-4fa4-ad4d-5254dc6681b2&folderid=de3f624f-2585-4cc7-92aa-2abd21c53878
- click into add in and press CTRL+V
- login with credentials (will send you in private message)

**Additional context**
When WebView2 ist uninstalled it works again!

[AB#32870744](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/32870744)

**Comment by ndeeH** [comment]
Any news - our Add Ins will fail for all existing customers !

**Comment by Fradulentus** [comment]
Рекомендую некоторое время не использовать данные инструменты для решения поставленных задач.
Хотите продолжить обсуждение в почте?

**Comment by champnic** [comment]
Hey @ndeeH - I believe the Office team has a separate process for bugs in their add-in system. I'll reach out to them and get back to you shortly.

**Comment by champnic** [comment]
@ndeeH I don't think I've received the credentials yet. You can send me an email at champnic@microsoft.com.

**Comment by ndeeH** [comment]
The content add in should look like this in powerpoint after inserting the report link:
![image](https://user-images.githubusercontent.com/11518242/115597505-a08b0380-a2d9-11eb-96ee-2713621ba3b7.png)

**Comment by champnic** [comment]
@ndeeH Do you know if your add-in worked properly with any version of WebView2? Or has it only been verified working correctly with the older WebView?

Do you know if your code has any differentiation based on web platform? WebView2 is Edge/Chromium based, whereas the older WebView is EdgeHTML based.

**Comment by ndeeH** [comment]
@champnic 
Hi Nic, 
- no we did not test especially against WebView2 versions. We simply saw this error since WebView2 was installed by Office/OS (?). The time before it worked using the EdgeHTML. Going back to EdgeHTML webview solved the problem (uninstall WebView2).

- We have only a differentiation against Chrome Browser (not WebView2) for the online version of of powerpoint.
There Chrome Browser did not accept the developer certificates and so we simply **skip** the backend version check when using chrome (see Home.js line 253). In the preview.idl.eu installation all certificates are valid.

We are using the **dialog API** in the content add in for the [authentication flow ](https://docs.microsoft.com/de-de/office/dev/add-ins/develop/auth-with-office-dialog-api):
1. our add in redirects using the dialog to login at the current backend (ThinkTecture)
2. The token cookie for this login is stored and the login redirects back to close the dialog
3. the add in now displays the content from the backend using an iFrame which should use the cookie stored before. 
(Therefore the content iFrame did not need to authenticate anymore)

This flow is for all browsers the identical.

_Is there any difference in handling cookies concerning dialog api or iFrames?_

**Comment by champnic** [comment]
@ndeeH
Hey Andi,

We believe the possible cause is because the cookie doesn't have the SameSite attribute, which on EdgeHTML defaults to SameSite=None, and on Chromium/Edge defaults to SameSite=Lax. This means the cookie doesn't get sent to the iframe. Could you try explicitly adding SameSite=None (docs suggest you will need 'secure' as well) on the cookies you need?
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite

Thanks!

**Comment by ndeeH** [comment]
@champnic 
Hi Nic,
bingo - you got it - the cookie coming from IdentityServerV3 was never set - cause there is no SameSite property set.
I was on the same track - here a screenshot I made at the time of the last post:
![image](https://user-images.githubusercontent.com/11518242/116592969-e3278e00-a920-11eb-8d1c-cb65e65bf381.png)

I disabled these flags in Chrome:
chrome://flags/#same-site-by-default-cookies
chrome://flags/#cookies-without-same-site-must-be-secure
and the Online Version of Powerpoint worked again.

Now I adopted IdentityServerV3 to allow SameSite=None when the current user agent would allow this (in V4 this is done like [this](https://github.com/IdentityServer/IdentityServer4/blob/45f8cb5d2215263528b6481ad9571729d7090c96/src/IdentityServer4/host/Extensions/SameSiteHandlingExtensions.cs)).

**Now - the Desktop Versions with WebView2 Runtime installed also work as expected.**

Do you know if there are experimental flags could be set in webview2 to temporary allow customers to work with the actual backend?

Bye and thank you, Andi

Here are some useful links:
https://leastprivilege.com/2020/01/05/identityserver3-and-upcoming-samesite-cookie-changes-in-browsers/
https://www.thinktecture.com/de/identity/samesite/prepare-your-identityserver/

**Comment by champnic** [comment]
I'm glad you've got the issue fixed!

Unfortunately there don't seem to be flags in Edge to enable the same behavior. We're planning on building such a command-line flag to allow app devs (like Office for add-ins) to keep the old default behavior and enable more seamless migration from EdgeHTML (no SameSite is the same as SameSite=None;Secure).

**Comment by ndeeH** [comment]
@champnic 
Hi Nic, 
the fix - setting "SameSite=None" everywhere - has implications to our normal web backend and normal logouts do not work anymore. 😣
Is there a chance to recognize the client is a MS Edge WebView2 Runtime? 
The UserAgent-String does not have any difference to MS Edge.
Thank you, Andi

**Comment by champnic** [comment]
@ndeeH 
When you say "normal web backend" does that also support a regular website and not just the Office add-in? If so, how are you currently handling cookies and login/logout on Chromium-based browsers?

We have a UserAgentString API but the app (Office, in this case) would have to set it. The cookie behavior in Edge and WebView2 should match, and only differ from EdgeHTML, which does have a different UA string. Is there a reason you need to differentiate between Edge and WebView2, and not just Edge/WebView2 and EdgeHTML?

**Comment by ndeeH** [comment]
@champnic 
Hi Nic,
setting **SameSite=None** resulted in problems when "deleting" the authentication cookie.
So I searched for a way to only set this mode when having a WebView2 as client.

BUT - now I found a solution for the logout problem: https://www.thinktecture.com/de/identity/samesite/how-to-delete-samesite-cookies/
TL;DR We had to set Secure=true, when deleting a cookie (= setting as expired).

With this changes the login and logouts to our website (report viewer) work in all chromium based browsers as expected.

**Background**
The Add In uses the website (report viewer) in embedded mode in an iFrame - so that's why we got the SameSite=Lax problem.
When not in an iFrame we never had those SameSite problems.

Thank you for your support!
